N°4261 Refactor ExceptionLog

piRGoif opened the pull request:
N°4261 was a way to : 

* Log more info on user portal exceptions (this is done in \Combodo\iTop\Portal\EventListener\ExceptionListener::onKernelException) : before we only logged the exception message using IssueLog
* Allow to log as EventIssue objects so that we can query those. Default level to log as EventIssue objects must be ERROR
* Configuration is done by class hierarchy. For example, defining Trace level for CoreException hierarchy (this class and all of its children)
* When logging, the channel used is the one defined in the configuration. The exception class is logged but in aContext data.

Doing a code review with Bruno, we agreed to do some little refactoring : 

* Level per exception class
  - Before the whole ExceptionLog::Log method was a total rewrite of its parent, with some code duplicates... not a good idea : we should better improve LogAPI to make other similar uses possible in the future !
  - The logic to get level from config must be in a GetMinLogLevel override
* Write to DB
  - Pull up this functionnality in LogAPI
  - Add a sCode parameter in GetLevelDefault

Doing this refactoring, I also improved : 

* Test the attributes set when creating the EventIssue object : during my dev I had crashes because I didn't filled all the mandatory fields... Having a PHPUnit test checking this will prevent future bugs to happen if attributes are modified in the class or in the object creation method
* Use Throwable instead of Exception : this was added in PHP 7.0 and will allow to catch both Exception and Error
* Because we need to have 2 statements on the same line in \Combodo\iTop\Test\UnitTest\Core\Log\ExceptionLogTest::testLogInFile, I modified the editorConfig file to allow disabling the formatter using comments.

piRGoif committed:
Mini modifications after review

piRGoif committed:
:recycle: Refactor ExceptionLog and LogAPI to use dtandard methods and overrides
* moved write to DB functionality from Exceptionlog to LogAPI
* overload EventIssue init in ExceptionLog
* split \LogAPI::GetMinLogLevel in sub functions to allow ExceptionLog override
* ExceptionLog::GetMinLogLevel impl to seek config on each class in the hierarchy

bruno-ds commented on core/log.class.inc.php:
```diff
@@ -645,22 +645,26 @@ public static function Log($sLevel, $sMessage, $sChannel = null, $aContext = arr
-		if (!static::IsLogLevelEnabled($sLevel, $sChannel)) {
-			return;
+		if (
+			(false === is_null(static::$m_oFileLog))
```
Since you're using a strict comparison, I propose to use `null !== static::$m_oFileLog` 
Or if you prefer to employ `is_null`, since it always return a bool: `! is_null(static::$m_oFileLog)` 

PS: This is not an important topic, but hey, I am a little obsessed...

piRGoif replied:
I like your first proposition !
About the second, we avoid it in the dev team as it's easy to miss the "!", whereas "false ===" is clear and can't be missed.

piRGoif replied:
Done in 31ae5f3a

bruno-ds commented on core/log.class.inc.php:
```diff
@@ -645,22 +645,26 @@ public static function Log($sLevel, $sMessage, $sChannel = null, $aContext = arr
 	 * @uses GetMinLogLevel
 	 */
-	final public static function IsLogLevelEnabled(string $sLevel, string $sChannel, string $sCode = 'log_level_min'): bool
+	final public static function IsLogLevelEnabled(string $sLevel, string $sChannel, string $sCode = self::ENUM_CONFIG_PARAM_FILE): bool
```
This is off topic, bu I think taht the `final`  keyword should be removed, since it made the initial implementation of ExceptionLog really cumbersome !

piRGoif replied:
I added the final keyword from the very start, when I added the method in f6d9d0f
As this method is an utility one that is calling really \LogAPI::GetMinLogLevel, I thought there shouldn't be any necessity to override it.
Do you have any use case in mind ?

piRGoif replied:
Discussed with Bruno, I'm closing this discussion.

bruno-ds replied:
We debated this with @piRGoif , my suggestion was wrong, sorry :(

bruno-ds commented on core/log.class.inc.php:
```diff
@@ -719,18 +740,63 @@ protected static function GetMinLogLevel($sChannel, $sCode = 'log_level_min')
+
+	protected static function GetMinLogLevelFromDefault($sChannel, $sCode = self::ENUM_CONFIG_PARAM_FILE)
+	{
 		if (isset($sLogLevelMin[static::CHANNEL_DEFAULT])) {
```
`$sLogLevelMin` isn't initialised, or did I missed something?

piRGoif replied:
Wooops, I'll have to take a look...

piRGoif replied:
Should be fixed by now, I'w solving this conversation

bruno-ds commented on core/log.class.inc.php:
```diff
@@ -719,18 +740,63 @@ protected static function GetMinLogLevel($sChannel, $sCode = 'log_level_min')
+		return null;
+	}
+
+	protected static function GetMinLogLevelFromDefault($sChannel, $sCode = self::ENUM_CONFIG_PARAM_FILE)
```
I don't really understand why you did broke the previous code into two separate functions since both have many branches that lead to "default" return values.

piRGoif replied:
Reviewed with Bruno directly, closing this conversation

bruno-ds commented on core/log.class.inc.php:
```diff
@@ -1169,110 +1235,72 @@ public static function LogException(Exception $oException, $aContext = array(),
-		}
-
+		$sConfiguredLevelForChannel = null;
+		$sExceptionClassInHierarchy = $sExceptionClass;
```
I thought that you were going to move this logic into `IssueLog ` did I misunderstood your intentions?

piRGoif replied:
Reviewed with Bruno directly, closing this conversation

piRGoif commented:
Rebased branch

piRGoif committed:
Change test
Thanks @bruno-ds !

piRGoif committed:
Fix LogAPI refactoring
* split of \LogAPI::GetMinLogLevel
* pull up of the EventIssue generation

piRGoif committed:
Fix log level for DB write

piRGoif committed:
Move tests for log in dedicated directory

piRGoif committed:
Fix LogAPITest
Now $oConfig->Get() is called for both log config (standard and DB)

piRGoif committed:
Fix ExceptionLog logging with wrong channel

piRGoif committed:
:wrench: EditorConfig : allow formatter:off annotation

piRGoif committed:
Fix Exception object logging
Those modifications made by Bruno were lost, certainly while rebasing
COnsidering the current code on develop :
* the Exception object is logged in $aContext
* when writing to file we are removing this object as it is too verbose

piRGoif commented:
Tests for log are back to OK now \o/
But... TransactionsTest throws an error, I need to solve this.

Also I'm going to write some info in the description to explain the idea behind this PR !

piRGoif committed:
:bulb: Fix comments

piRGoif committed:
Fix EventIssue generation
* there were some mandatory fields not fixed :/
* if we failed during DB write, we tried db write again by calling IssueLog !
* use the parent method in ExceptionLog::GetEventIssue

piRGoif committed:
Fix line pushed by mistake

piRGoif committed:
Test for EventIssue generation
We must define all mandatory attr !

piRGoif committed:
Use Throwable instead of Exception
This way we will be able to log both Exception and Error
Throwable was added in PHP 7.0 so we can defenitively use it in this version
See https://www.php.net/manual/fr/class.throwable.php

piRGoif commented:
2 more commits this morning : 
* use Throwable instead of Exception in \ExceptionLog::LogException signature
* test that \LogAPI::GetEventIssue really sets all the mandatory fields in the object

The PR description was updated accordingly !

piRGoif committed:
Fix deprecated notices in LogAPITest

piRGoif committed:
Fix LogAPITest : class not found MetaModel

bruno-ds commented on core/log.class.inc.php:
```diff
@@ -645,22 +654,27 @@ public static function Log($sLevel, $sMessage, $sChannel = null, $aContext = arr
-			return;
+		if (
+			(null !== static::$m_oFileLog)
+			&& static::IsLogLevelEnabled($sLevel, $sChannel, static::ENUM_CONFIG_PARAM_FILE)
```
about `static::ENUM_CONFIG_PARAM_FILE`: For sure, using constants provide robustness, and it's not an error to use them, but personally I'm really not fond of them for a readability POV.

For instance, an alternative would have been to implement  `IsLogLevelEnabledForFiles` and `IsLogLevelEnabledForDb` (the names are lame, but I like the idea).

bruno-ds commented on core/log.class.inc.php:
```diff
@@ -645,22 +654,27 @@ public static function Log($sLevel, $sMessage, $sChannel = null, $aContext = arr
+		$sMinLogLevel = self::GetMinLogLevel($sChannel, $sLogConfigKey);
 
-		if ($sMinLogLevel === false || $sMinLogLevel === 'false') {
+		// the is_bool call is to remove a IDE O:) warning as $sMinLogLevel is typed as string
```
I'm not even sure the `$sMinLogLevel === 'false'` is required, maybe I coded it preventatively (because I don't know well how iTop enforce type correctness in the config)

piRGoif replied:
Reviewed with Bruno directly, closing this conversation

bruno-ds commented on core/log.class.inc.php:
```diff
@@ -719,16 +756,65 @@ protected static function GetMinLogLevel($sChannel, $sCode = 'log_level_min')
+			// calling low level methods : if we would call Error() for example we would try to write to DB again...
+			static::$m_oFileLog->Error('Failed to log issue into the DB', LogChannels::CORE, [
+				'exception message' => $e->getMessage(),
+				'exception stack'   => $e->getTraceAsString(),
```
Writing the call stack is a hard choice to make, I think this should be debated, maybe with molkobain or eric?
(I think it's a good idea, but fear an stratospheric pollution of the logs in case of certain circumstances such as read_only mode)

bruno-ds commented on core/log.class.inc.php:
```diff
@@ -719,16 +756,65 @@ protected static function GetMinLogLevel($sChannel, $sCode = 'log_level_min')
+		}
+
+		try {
+			self::$oLastEventIssue = static::GetEventIssue($sMessage, $sChannel, $aContext);
```
extracting the instantiation really improve the readability, congrats!

piRGoif replied:
And also improves extensibility :) Both were the goals to reach here O:)

bruno-ds commented on core/log.class.inc.php:
```diff
@@ -719,16 +756,65 @@ protected static function GetMinLogLevel($sChannel, $sCode = 'log_level_min')
+		$oEventIssue->Set('issue', $sMessage);
+		$oEventIssue->Set('message', $sMessage);
+		$oEventIssue->Set('date', $sDate);
+		$oEventIssue->Set('userinfo', UserRights::GetUserFriendlyName());
```
This may result in additional queries to the database, I'm a little afraid by the impact it mayu have for corner cases.

piRGoif replied:
Reviewed with Bruno directly, closing this conversation

bruno-ds commented on core/log.class.inc.php:
```diff
@@ -744,15 +830,26 @@ protected static function GetConfig(): ?Config
-	protected static function GetLevelDefault(): string
+	protected static function GetLevelDefault(string $sLogConfigKey): string
 	{
+		if ($sLogConfigKey === self::ENUM_CONFIG_PARAM_DB) {
```
while totally producing the same result, maybe a switch would improve the readability, or may just a 2d if `if ($sLogConfigKey === self::ENUM_CONFIG_PARAM_FILE) {` .

(I'm not 100% percent sure it would be better, I just think that the current code do not explcitely expose the 3 handled cases with juste one if)

piRGoif replied:
Reviewed with Bruno directly, closing this conversation

bruno-ds commented on core/log.class.inc.php:
```diff
@@ -965,8 +1063,12 @@ public static function DeprecatedNoticesErrorHandler(int $errno, string $errstr,
-	protected static function GetLevelDefault(): string
+	protected static function GetLevelDefault(string $sLogConfigKey): string
 	{
+		if ($sLogConfigKey === self::ENUM_CONFIG_PARAM_DB) {
```
I'm surprised that `utils::IsDevelopmentEnvironment()` has impact only for the file backend and not the db backend.

piRGoif replied:
Reviewed with Bruno directly, closing this conversation

bruno-ds commented on core/log.class.inc.php:
```diff
@@ -1133,147 +1235,171 @@ private function GetLogFileNameBuilderClassName()
+	 * The parameter order is not standard, but in our use case, the resulting API is way more convenient this way !
 	 */
-	public static function LogException(Exception $oException, $aContext = array(), $sLevel = self::LEVEL_WARNING)
+	public static function LogException(Throwable $oException, $aContext = array(), $sLevel = self::LEVEL_WARNING): void
```
Thanks for the change for `Throwable` type hint!

bruno-ds commented on core/log.class.inc.php:
```diff
@@ -1133,147 +1235,171 @@ private function GetLogFileNameBuilderClassName()
+
+		$sExceptionClass = get_class($oException);
+
 		if (empty($aContext[self::CONTEXT_EXCEPTION])) {
```
I am the faulty one on this one again!
=> using array_merge() would treat all 4 array keys at once, and, I think, improve overall readability. 

ie: 
```
$aDefaultContext = [
    self::CONTEXT_EXCEPTION => $oException,
    'exception class' => $sExceptionClass, 
    'file' => $oException->getFile(),
    'line' => $oException->getLine(),
]
$aContext = array_merge($aDefaultContext, $aContext);
```

bruno-ds commented on core/log.class.inc.php:
```diff
@@ -1133,147 +1235,171 @@ private function GetLogFileNameBuilderClassName()
+	 *
+	 * @return string|null the current or parent class name defined in the config, otherwise null if no class of the hierarchy found in the config
+	 */
+	protected static function GetExceptionClassInConfig($sExceptionClass, $sLogConfigKey = self::ENUM_CONFIG_PARAM_FILE)
```
1/ I'm not fond of the method  name as I find it hard to understand.
Maybe replace `Get` by `Find` as the result is computed, but also the class hierarchy traversal is not really guessable just by the method name.

2/ Maybe this method would benefit from being moved into it's parent, as the use case may be implemented by other classes than ExceptionLog?

piRGoif replied:
Reviewed with Bruno directly, closing this conversation

bruno-ds reviewed (commented):
I found it really hard to read it with PHPstorm's PR review UI, maybe we should read the code together and exchange oraly.

piRGoif committed:
Rename method
Thanks @brundo-ds !

piRGoif committed:
Rename method in test also (wooops)

piRGoif commented:
We did a face to face review with @bruno-ds :)

Here are the conclusions / TODO : 

- [x] LogAPITest : there was a line removed in a dataprovider, why ? 
- LogAPI 
  - [x] rename $sLogConfigKey parameters to $sConfigKey
  - [x] WriteToDb : add a static attribute to avoid re-entrance (similar to what is done in \DBObject::DBUpdate)
  - [x] GetEventIssue : \UserRights::GetUserFriendlyName is generating queries to the DB, we should find another better way to print the current user info (Session::Get('auth_user'); that is set in \LoginWebPage::OnLoginSuccess or \UserRights::$m_oUser maybe ?)
  - [x] GetLevelDefault : to improve readability use switch or if instead
- [x] \DeprecatedCallsLog::GetLevelDefault : add a PHPDoc to explain the behavior
- Exceptionlog
  - [ ] maybe pull up in LogAPI the search in the class hierarchy ? (this may be used when logging inline image as we're using the DM class as channel...)
  - [ ] the exception is stored in aContext and this is bad :/ We avoided an extra param but this isn't nice. Maybe change the parameter to have 2 root keys : data (current content) and metadata (the exception instance we need to keep) ?
  - [x] in the unit test use a Throwable / Error
  - [x] Log() : instead in LogAPI extract write operations in a dedicated method and override only this one
  - [ ] ExceptionClassFromHierarchy : to improve readability extract to a dedicated method

piRGoif committed:
Fix line removed in \Combodo\iTop\Test\UnitTest\Core\Log\LogAPITest::LogApiProvider

piRGoif commented:
Removal of a dataprovider line : in \Combodo\iTop\Test\UnitTest\Core\Log\LogAPITest::LogApiProvider
To show it, use the git history, open diff between current version and the one in 3abcd59b0369480faf8323f528d4e4503475d989 :

![image](https://user-images.githubusercontent.com/8947448/137882519-f99ce71f-2e61-424c-9a7d-2373c51a5e4e.png)

This line was causing the test to fail, because the current test impl uses mock for the config and so we had : 
`TypeError : Return value of LogAPI::GetConfig() must be an instance of Config or null, array returned`

I replaced the `array()` by null : OK.
Committing this ! 357254f2

piRGoif committed:
Improve \DeprecatedCallsLog::GetLevelDefault
- add phpdoc
- call parent method instead of using constants directly (this is the responsibility of the parent method !)

piRGoif committed:
Rename sLogConfigKey parameter to sConfigKey

piRGoif committed:
Rewrote \LogAPI::GetLevelDefault with switch to improve readability

piRGoif committed:
Protect WriteToDb against reentrence

piRGoif committed:
Refactor to be able to override only the log writing (and not the whole Log() method)
Also implements ExceptionLog::Log to throw an exception (we don't want this method to be used !)

bruno-ds commented on core/log.class.inc.php:
```diff
@@ -782,6 +782,14 @@ protected static function WriteToDb(string $sMessage, string $sChannel, array $a
 		}
 
+		// Protect against reentrance
+		static $aWriteToDbReentrance = array();
```
I think that a boolean would have been sufficient as it would have mean
> while you're writing a log to DB, then the code inside this write cannot write a log to db

your implementation mean:
> while you're writing a log to DB, then you are forbidden to write the exact same message to db

piRGoif replied:
Indeed you're right !

piRGoif replied:
Fixed in 4cc77608

piRGoif committed:
User info for EventIssue + use new method in FileLog

piRGoif committed:
Fix reentrance flag in \LogAPI::WriteToDb

piRGoif committed:
Fix CI (woops)

piRGoif committed:
Test Throwable signature
